enhancements/cluster-high-availability-mode-api.md

```diff
@@ -0,0 +1,360 @@
+---
+title: cluster-high-availability-mode-api
+authors:
+  - "@dhellmann"
+  - "@eranco"
+  - "@mrunalp"
+  - "@romfreiman"
+reviewers:
+  - TBD
+  - "@markmc"
+  - "@deads2k"
+  - "@wking"
+  - "@hexfusion"
+approvers:
+  - TBD
+  - "@eparis"
+  - "@derekwaynecarr"
+creation-date: 2020-12-02
+last-updated: 2020-12-02
+status: implementable
+see-also:
+  - "/enhancements/single-node-production-edge-cluster-profile.md"
+  - "/enhancements/update/cluster-profiles.md"
+---
+
+# Cluster High-availability Mode API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+As we add more deployment topologies for OpenShift, we need a way to
+communicate expectations to the operators managing the cluster. This
+enhancement describes an API for telling operators how hard they
+should work to provide high-availability.
+
+## Motivation
+
+We want to avoid having cluster profiles use different manifests to
+configure operators managed by the `cluster-version-operator`, and
+instead rely on in-cluster APIs. This approach will make the same
+information available to operators managed by either the
+`cluster-version-operator` or `operator-lifecycle-manager`.
+
+We do not want to expose the cluster profile name directly to most of
+the operators, because we do not want to have to update operators each
+time we add a new profile.
+
+We want the setting to be global to the cluster, rather than using
+operator-specific APIs, because that simplifies the integration with
+the installer and allows it to be applied consistently in all
+operators that need them.
+
+### Goals
+
+- Expose high-availability expectations to operators running in a
+  cluster so they can configure their operands appropriately.
+
+### Non-Goals
+
+- Expose cluster profile names to operators.
+- Provide a mutable API for changing the high-availability mode (at
+  least for now).
+- Predict other capability settings that we might need to expose.
+- Define a more generic "capabilities" API.
+
+## Proposal
+
+### Cluster Operators
+
+The `infrastructure.config.openshift.io` [informational
+resource](https://docs.openshift.com/container-platform/4.6/installing/install_config/customizations.html#informational-resources_customizations),
+is expanded to include a new field to describe the high-availability
+mode desired for cluster operators running on the cluster.
+
+The `highAvailabilityMode` field will use an enum to express the
```

Comment:
I'm not quite sure I understand how splitting the profile into two fields, one which seems to be bool (it's HA or not, there's no other option) is unnecessarily introduced as enum, unless I'm missing some more argument what different HA mode we're planning, we'll help here. 
Either way, profile or capabilities, operators will need to know how and what to do with their operands based on either of the fields. Why than overcomplicate this API by adding those extra layers?

Reply:
Only one field is proposed - the HA mode field

It's an enum because a previous reviewer (Trevor, Clayton?) recommended never using a bool field because it can't be expanded in the future

Maybe you've interpreted this as meaning there will also be a "cluster profile" field for operators to react to? (Just guessing). While cluster profiles are mentioned above, the intention AIUI is that the HA mode field is the only field that operators should react to.

Reply:
Fair, I think I misread the capabilities from alternatives :+1:

Reply:
Although, honestly this approach is going towards profiles more than just `haMode`. I'm pretty sure it's a matter of time when this is how this will be interpreted. Why not going then with a profile, deploymentMode, or something that is describing the what the cluster will look like, rather than focusing only on HA.

Reply:
We're not using a profile-name-based API because we don't want to add a bunch of profiles. Defining and maintaining them is a lot of work.

Adding more focused APIs means that only the operators that need to respond to those APIs have to be updated, and (depending on what the API does) they don't necessarily need to be updated all at the same time.

Reply:
Personally I think we're bikeshedding this particular field a bit too much - yes, we'll need to maintain it ~forever and maybe later we could end up wanting something more sophisticated but... in the end, IMO by *far* the largest maintenance burden isn't going to be these API fields, it's going to be all the work in the operators to honor them and the CI jobs to test different variants, etc.

Whether it's a bool or a string or a list of profiles/flags is really just a small tip of the iceberg - IMO let's just be sure it's not *wrong* of course and then move onto thinking about all those harder problems :wink:

Reply:
We don't allow booleans generally in long term apis (and we never have), see Kube API conventions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types.  No one should add booleans to spec fields without the belief we would never have another mode.  And I don't think even from the discussion so far that we've established that.

If we want to be precise, we're trying to target a specific behavior "single node" which is NOT the same as "single master, many workers" (this configuration could very well want ingress controllers to be highly available).  So we should be explicit about what the user gets, and define a policy that is scoped to what we want, and if we come up with use cases in the future.  I don't think `haMode` is going to be the best name here - if we're being precise (and we should be!), we're describing the preferred cluster topology for the control plane and nothing more, so I would suggest exploring naming towards things like:

`controlPlaneTopology` (topology is a bit more specific than policy, but there may be other words) with values 

* `Singleton` or `SingleNode` or `Minimal` (no strong preference, but each of these is describing the desired topology)
* `HighlyAvailable` (the current setting, which attempts to be highly available out of the box)

We already support additional nodes on HighlyAvailable clusters, and scheduling policy is orthogonal to the setting (these are good things).  If we in the future decide to allow node addition to single node control plane, we don't have to change the constant because we're describing the control plane topology.  If we decide that we need to embody in the product a built-in failover mode (vs having it layered on top), then a new enum value would be `ActivePassive` or `MinimalFailover` or `SingleNodeFailover`.  

It's always better to describe using the enum the thing you expect to see at the outcome as the user would see it.  We know other flags will control other aspects.  That's partially why I don't like `None` - it doesn't describe the current state, whereas a more specific enum value and a more policy focused field name is part of our api approach.

Reply:
`ActivePassive`, `MinimalFailover`, and `SingleNodeFailover` all sound more like HA modes than topologies to me. We could call the field `haPolicy` I guess?

What sorts of changes would we anticipate having to make to our existing operators to support those new values?

If we want to focus on the topology, would it be better to reflect the `ControlPlane.Replicas` value from the install-config as a number instead of using an enum? And then deal with HA behaviors with an additional field later? I could, for example, see use in a configuration with 3 etcd replicas inside the cluster and HA for routing or API services handled outside of the cluster.

Reply:
I like `controlPlaneTopology: SingleNode` personally.

>  If we in the future decide to allow node addition to single node control plane,

That...hurts my head.

Reply:
knowing the topology would provide a nice hint for our web console to not do things that have been problematic when there are no control plane hosts visible in cluster.

Reply:
So maybe the following:

ControlPlaneTopology: Standalone | External
ControlPlaneAvailability: HA | Minimal

above would be used by things like etcd-operator, auth-operator, stuff that expect to never run on role=worker.


InfrastructureAvailability: HA | Minimal

covers components that target role=worker or infra

This would let us support an IBM Cloud Satellite cluster with 1 worker whose control plane is externalized by not telling the router to expect HA, or the monitoring stack should reduce its footprint AND support the standalone topology needs for single-node or ha.

Reply:
oh, external.  Hrm.  If it's external, you don't get to say what the availability is inside the cluster? CPA seems irrelevant for external.

Reply:
I'm not against InfrastructureAvailability but I hate infrastructure as a word generally, and availability is wierd.  Hrm...

```
controlPlaneTopology: SingleNode|HighlyAvailable|External
infrastructureTopology: Minimal|HighlyAvailable
```

Reply:
Does that mean we want to combine topology and availability into 1 field after all? Is there an "external" mode for the infrastructure services, too?

Reply:
(derek and i couldn't think of a better word than infrastructure in slack either, and controlPlane and infrastructure are words we use elsewhere so it would be consistent to use them here).  Also in the future we could have external infrastructure but it's unlikely ALL of it would be external

Reply:
OK, so 2 enum fields, and for now only the control plane supports external.

Do we need to add the external value as part of the first implementation, or can we leave that for a follow-up enhancement for the team working on the external control plane?

Reply:
What would External/External mean?

What's the distinction between control plane and infrastructure conceptually when using SingleNode/SingleNode?

Reply:
We know there is a possibility (unlikely, let's say 10-40%) that we may in the future want to support single node control planes but multiple infra.  Likewise, we have a use case for External|SingleNode.  We will have a use case for External|HighlyAvailable.  We probably won't have a biz use case for HighlyAvailable|SingleNode, but occasionally people set up small test environments (compact has mostly removed the need for this and i suspect hypershift will remove the rest of the need for it in the future, i.e. External|SingleNode).  External|External might be used (hypothetically) by third party operators that have an integration with Prometheus to directly access data in alertmanager via pod exec (i.e. using infrastructure access rather than api access, which we would generally frown on anyway, but eh).

External would be a followup whenever that relevant feature lands.  You may want to describe "potential future values of this flag" here so people understand the space you are creating and we can refer back to it in later enhancements.  You might mention here that "a future revision may desire to describe an active passive cluster, in which case controlPlaneTopology might acquire a new enumeration `ActivePassive` if such functionality is required"

Reply:
@smarterclayton Where can I learn more about 'compact'?

Reply:
https://github.com/openshift/enhancements/blob/93efdd811027688182575aec7baaa476279705c3/enhancements/compact-clusters.md

Which after this merges will need to be updated to reference what we agree on here

Reply:
The latest updates to the text have broken the link to this thread, so I am going to resolve it. We can continue the discussion with a new thread that GitHub can show in the right place on the text.